pkg/controller/seed-controller-manager/etcdbackup/etcd_backup_controller.go

```diff
@@ -0,0 +1,983 @@
+	if err := r.ensureEtcdClientSecret(ctx, cluster); err != nil {
+		return nil, errors.Wrap(err, "failed to create backup secret")
+	}
+
+	var nextReconcile, totalReconcile *reconcile.Result
+	var err error
+	errorReconcile := &reconcile.Result{RequeueAfter: 1 * time.Minute}
+
+	if nextReconcile, err = r.ensureNextBackupIsScheduled(ctx, backupConfig, cluster); err != nil {
+		return errorReconcile, errors.Wrap(err, "failed to ensure next backup is scheduled")
+	}
+
+	totalReconcile = minReconcile(totalReconcile, nextReconcile)
+
+	if nextReconcile, err = r.startPendingBackupJobs(ctx, backupConfig, cluster); err != nil {
+		return errorReconcile, errors.Wrap(err, "failed to start pending and update running backups")
+	}
+
+	totalReconcile = minReconcile(totalReconcile, nextReconcile)
+
+	if nextReconcile, err = r.startPendingBackupDeleteJobs(ctx, backupConfig, cluster); err != nil {
+		return errorReconcile, errors.Wrap(err, "failed to start pending backup delete jobs")
+	}
+
+	totalReconcile = minReconcile(totalReconcile, nextReconcile)
+
+	if nextReconcile, err = r.updateRunningBackupDeleteJobs(ctx, backupConfig, cluster); err != nil {
+		return errorReconcile, errors.Wrap(err, "failed to update running backup delete jobs")
+	}
+
+	totalReconcile = minReconcile(totalReconcile, nextReconcile)
+
+	if nextReconcile, err = r.deleteFinishedBackupJobs(ctx, backupConfig, cluster); err != nil {
+		return errorReconcile, errors.Wrap(err, "failed to delete finished backup jobs")
+	}
+
+	totalReconcile = minReconcile(totalReconcile, nextReconcile)
+
+	if nextReconcile, err = r.handleFinalization(ctx, backupConfig, cluster); err != nil {
+		return errorReconcile, errors.Wrap(err, "failed to delete finished backup jobs")
+	}
+
+	totalReconcile = minReconcile(totalReconcile, nextReconcile)
+
+	return totalReconcile, nil
+}
+
+func minReconcile(reconciles ...*reconcile.Result) *reconcile.Result {
+	var result *reconcile.Result
+	for _, r := range reconciles {
+		if result == nil || (r != nil && r.RequeueAfter < result.RequeueAfter) {
+			result = r
+		}
+	}
+	return result
+}
+
+// ensure a backup is scheduled for the next time slot after the current time, according to the backup config's schedule.
+// "schedule a backup" means set the scheduled time, backup name and job names of the corresponding element of backupConfig.Status.CurrentBackups
+func (r *Reconciler) ensureNextBackupIsScheduled(ctx context.Context, backupConfig *kubermaticv1.EtcdBackupConfig, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	if backupConfig.DeletionTimestamp != nil || cluster.DeletionTimestamp != nil {
+		// backupConfig is deleting. Don't schedule any new backups.
+		return nil, nil
+	}
+
+	var backupToSchedule *kubermaticv1.BackupStatus
+
+	if len(backupConfig.Status.CurrentBackups) > 2*backupConfig.GetKeptBackupsCount() {
```

Comment:
Yes.

Basically delete jobs are the only things that know how to delete a backup, so if they fail, we can't be sure if there are remnants left behind in the object store. So I'm retrying delete jobs until they succeed, and only then remove the backup entry from .Status.CurrentBackups. So if they keep failing (e.g. because there's something wrong with the seed cluster), entries pile up in .Status.CurrentBackups, so at some point I stop scheduling new backups and set a condition ("SchedulingActive") to false on the backupconfig to indicate this. Everything will recover automatically once the delete jobs start succeeding. I basically wanted to avoid a situation where an administrator would have to clean up stuff manually in the object store after a problem with the seed cluster prevented jobs from being created for a while.